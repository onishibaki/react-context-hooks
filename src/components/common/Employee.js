import React, { useState } from "react";
import { useEmployeeState } from "../context";
import styled from "styled-components";

import { ActionTypes, addEmployee, updateEmployee } from "../action";

const EmployeeContainer = styled.div`
  border: 1px solid black;
  background: rgba(0, 0, 0, 0.5);
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
  text-align: center;
`;

const EmployeeWrapper = styled.div`
  border: 1px solid rgb(17, 17, 17);
  background-color: white;
  margin-top: 158px;
  display: inline-block;
  padding: 35px 10px 10px 10px;
  position: relative;
`;
const Button = styled.button`
  text-decoration: none;
  border: 1px solid #000000;
  padding: 10px 20px;
  font-weight: 700;
  color: black;
  font-size: 12px;
  margin: 10px 0;
  display: block;
  &:hover {
    background: #00000059;
    color: white;
  }
`;
const SpanButton = styled.span`
  font-weight: 700;
  color: black;
  font-size: 15px;
  margin: 10px 0;
  display: inline-block;
  position: absolute;
  top: -8px;
  right: 9px;
  cursor: pointer;
}
`;

const Input = styled.input`
  margin: 0 10px;
}
`;

export const Employee = () => {
  //Use Context
  const [
    {
      form,
      employee: { firstName, lastName, employee_id: employeeId },
      formMessage,
    },
    dispatch,
  ] = useEmployeeState();

  //Set state for First Name and Last Name Changes
  const [stateFirstName, setStateFirstName] = useState("");
  const [stateLastName, setStateLastName] = useState("");

  //Set state for setting blank value
  const [stateUpdateFirstName, setStateUpdateFirstName] = useState(null);
  const [stateUpdateLastNamet, setStateUpdateLastName] = useState(null);

  //Set the First Value
  let firstNameParameter = firstName;
  let lastNameParameter = lastName;

  //Parameter for Save
  const employeeParamSave = {
    stateFirstName,
    stateLastName,
  };

  //Checking for State if First name and Last name is blank to throw error
  if (stateUpdateFirstName === "") {
    firstNameParameter = stateUpdateFirstName;
  }
  if (stateUpdateLastNamet === "") {
    lastNameParameter = stateUpdateLastNamet;
  }

  //Parameter for Update
  const employeeParamUpdate = {
    employeeId,
    stateFirstName: stateFirstName || firstNameParameter,
    stateLastName: stateLastName || lastNameParameter,
  };

  //Getting the value of Input changes
  const inputChangedHandlerFirstName = (event) => {
    setStateFirstName(event.target.value);
    setStateUpdateFirstName("");
  };

  const inputChangedHandlerLastName = (event) => {
    setStateLastName(event.target.value);
    setStateUpdateLastName("");
  };

  return (
    //Condition to display Save and Update Screen
    form !== "" &&
    form !== "view" && (
      <EmployeeContainer className="employee-container">
        <EmployeeWrapper>
          <SpanButton
            onClick={() => {
              dispatch({
                type: ActionTypes.SET_FORM_STATE,
                form: "",
                data: {},
              });
              dispatch({
                type: ActionTypes.FORM_MESSAGE,
                data: "",
              });
              setStateLastName("");
              setStateFirstName("");
              setStateUpdateFirstName(null);
              setStateUpdateLastName(null);
            }}
          >
            X
          </SpanButton>
          <label>First Name: </label>
          <Input
            name="fname"
            defaultValue={firstName}
            onChange={inputChangedHandlerFirstName}
          ></Input>
          <label>Last Name: </label>
          <Input
            name="lname"
            defaultValue={lastName}
            onChange={inputChangedHandlerLastName}
          ></Input>
          {Object.keys(formMessage).length !== 0 && (
            <h2>{formMessage.message}</h2>
          )}
          {form !== "update" ? (
            <Button
              onClick={() => {
                addEmployee(employeeParamSave).then(function (response) {
                  dispatch(response);
                });
              }}
            >
              Save
            </Button>
          ) : (
            <Button
              onClick={() => {
                updateEmployee(employeeParamUpdate).then(function (response) {
                  dispatch(response);
                });
              }}
            >
              Update
            </Button>
          )}
        </EmployeeWrapper>
      </EmployeeContainer>
    )
  );
};
