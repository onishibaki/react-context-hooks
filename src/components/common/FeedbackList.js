import React from "react";
import { useEmployeeState } from "../context";
import styled from "styled-components";

import { ActionTypes, updateEmployeeFlag } from "../action";
import { Feedback } from "./Feedback";

const FeedbackListContainer = styled.div`
  background: rgba(0, 0, 0, 0.5);
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
  text-align: center;
`;

const FeedbackListWrapper = styled.div`
  border: 1px solid rgb(17, 17, 17);
  background-color: white;
  display: inline-block;
  position: relative;
`;

const Button = styled.button`
  text-decoration: none;
  border: 1px solid #000000;
  padding: 10px 20px;
  font-weight: 700;
  color: black;
  font-size: 12px;
  margin: 25px 10px 10px 10px;
  display: inline-block;
  height: 100%;
  &:hover {
    background: #00000059;
    color: white;
  }
`;

const Unordered = styled.ul`
  list-style: none;
  padding: 0;
  display: flex;
  justify-content: space-around;
  flex-direction: column;
  padding: 0 20px;
  li {
    display: flex;
    margin: 5px 0;
  }
`;

const Paragraph = styled.p`
  border: 1px solid black;
  padding: 5px;
  margin-right: 20px;
  text-align: left;
  word-wrap: break-word;
  font-size: 12px;
  width: 500px;
  height: 54px;
`;

const SpanButton = styled.span`
  font-weight: 700;
  color: black;
  font-size: 15px;
  margin: 10px 0;
  display: inline-block;
  position: absolute;
  top: -8px;
  right: 9px;
  cursor: pointer;
}
`;

export const FeedBackList = () => {
  const [
    {
      form,
      employee: { employee_id, feedbacklist },
    },
    dispatch,
  ] = useEmployeeState();

  const feedBackParam = {
    employeeId: employee_id,
    flag: 1,
  };

  return (
    form === "view" && (
      <FeedbackListContainer className="feedback-list-container">
        <FeedbackListWrapper>
          <SpanButton
            onClick={() =>
              dispatch({
                type: ActionTypes.SET_FORM_STATE,
                form: "",
                data: {},
              })
            }
          >
            X
          </SpanButton>
          <Button
            onClick={() => {
              dispatch({
                type: ActionTypes.SET_FORM_STATE_FEEDBACK,
                form: "save",
                data: {},
                id: employee_id,
              });
            }}
          >
            Add Feedback
          </Button>
          <Button
            onClick={() => {
              updateEmployeeFlag(feedBackParam);
              dispatch({
                type: ActionTypes.SET_FORM_STATE,
                form: "",
                data: {},
              });
            }}
          >
            Assign Feedback
          </Button>
          <Unordered>
            {feedbacklist.map((feedbacklists) => {
              const { feedback_id, feedback } = feedbacklists;
              return feedback_id !== null ? (
                <li key={feedback_id}>
                  <Paragraph>{feedback}</Paragraph>
                  <Button
                    onClick={() => {
                      dispatch({
                        type: ActionTypes.SET_FORM_STATE_FEEDBACK,
                        form: "update",
                        data: feedbacklists,
                        id: employee_id,
                      });
                    }}
                  >
                    Update
                  </Button>
                </li>
              ) : (
                <h5 key="0">No Feedback</h5>
              );
            })}
          </Unordered>
          <Feedback />
        </FeedbackListWrapper>
      </FeedbackListContainer>
    )
  );
};
