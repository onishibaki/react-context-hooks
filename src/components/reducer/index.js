import React, { useReducer } from "react";

import { ActionTypes } from "../action";
import { EmployeesContext } from "../context";

const appStateReducer = (state, action) => {
  switch (action.type) {
    case ActionTypes.GET_ALL_EMPLOYEE:
      return {
        ...state,
        employees: action.data,
      };
    case ActionTypes.FORM_MESSAGE:
      return {
        ...state,
        formMessage: action.data,
      };
    case ActionTypes.SET_FORM_STATE:
      return {
        ...state,
        form: action.form,
        employee: action.data,
      };
    case ActionTypes.SET_FORM_STATE_FEEDBACK:
      return {
        ...state,
        formValue: action.form,
        feedbackValue: action.data,
        employeeId: action.id,
      };
    default:
      return state;
  }
};

const initialState = {
  employees: {},
  form: "",
  employee: {},
  formValue: "",
  feedbackValue: {},
  id: "",
  formMessage: "",
};

export const AppEmployeesProvider = ({ children }) => {
  const employeesList = useReducer(appStateReducer, initialState);
  return (
    <EmployeesContext.Provider value={employeesList}>
      {children}
    </EmployeesContext.Provider>
  );
};
