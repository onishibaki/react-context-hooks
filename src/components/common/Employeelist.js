import React from "react";
import styled from "styled-components";

import { ActionTypes, DeleteEmployee, GetAllEmployees } from "../action";

import { useEmployeeState } from "../context";
import { Employee } from "./Employee";
import { FeedBackList } from "./FeedbackList";

const EmployeesWrapper = styled.div`
  border: 1px solid #000000;
  margin-top: 30px;
`;
const EmployeesDetailWrapper = styled.div`
  width: 60%;
`;
const TopMenu = styled.div`
  margin: 22px 10px;
  a {
    text-decoration: none;
    border: 1px solid #000000;
    padding: 10px 20px;
    font-weight: 700;
    color: black;
    font-size: 12px;
    margin: 0 20px;
    &:hover {
      background: #00000059;
      color: white;
    }
  }
`;
const Button = styled.button`
  text-decoration: none;
  border: 1px solid #000000;
  padding: 10px 20px;
  font-weight: 700;
  color: black;
  font-size: 12px;
  margin: 0 20px;
  &:hover {
    background: #00000059;
    color: white;
  }
`;

const SaveButton = styled(Button)`
  float: right;
`;
const Unordered = styled.ul`
  list-style: none;
  padding: 0;
  display: flex;
  justify-content: space-around;
  flex-direction: column;
  padding: 0 20px;
  li {
    display: flex;
    margin: 5px 0;
  }
`;

const Span = styled.span`
  font-size: 15px;
  padding: 0 20px;
  font-weight: 700;
`;

export const Employeelist = ({ props }) => {
  const { employees, status, message } = props;
  // eslint-disable-next-line
  const [stateDispatch, dispatch] = useEmployeeState();
  GetAllEmployees().then(function (response) {
    dispatch(response);
  });

  return status === 0 ? (
    <div className="main-container">{message}</div>
  ) : (
    <div className="main-container">
      <EmployeesWrapper>
        <TopMenu>
          <SaveButton
            onClick={() => {
              dispatch({
                type: ActionTypes.SET_FORM_STATE,
                form: "save",
                data: {},
              });
            }}
          >
            Add
          </SaveButton>
          <span>
            <a href="/">Home</a>
          </span>
        </TopMenu>
        <Unordered>
          {employees.map((employee) => {
            const { employee_id, firstName, lastName } = employee;
            return (
              <li key={employee_id}>
                <EmployeesDetailWrapper>
                  <label>First name: </label>
                  <Span>{firstName}</Span>
                  <label>Last name: </label>
                  <Span>{lastName}</Span>
                </EmployeesDetailWrapper>
                <Button
                  onClick={() => {
                    DeleteEmployee(employee_id);
                  }}
                >
                  Delete
                </Button>
                <Button
                  onClick={() => {
                    dispatch({
                      type: ActionTypes.SET_FORM_STATE,
                      form: "update",
                      data: employee,
                    });
                  }}
                >
                  Update
                </Button>
                <Button
                  onClick={() => {
                    dispatch({
                      type: ActionTypes.SET_FORM_STATE,
                      form: "view",
                      data: employee,
                    });
                  }}
                >
                  View
                </Button>
              </li>
            );
          })}
        </Unordered>
      </EmployeesWrapper>
      <Employee />
      <FeedBackList />
    </div>
  );
};
