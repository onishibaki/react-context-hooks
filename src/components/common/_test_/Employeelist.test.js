import React from "react";
import { render, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import Adapter from "enzyme-adapter-react-16";
import { configure, mount } from "enzyme";

import { Employee } from "../Employee";
import { FeedBackList } from "../FeedbackList";
import { Employeelist } from "../Employeelist";
import { AppEmployeesProvider } from "../../reducer";

configure({ adapter: new Adapter() });

const employees = {
  status: 1,
  message: "Success",
  employees: [
    {
      employee_id: 1,
      firstName: "Mau",
      lastName: "Maya",
      flag: "1",
      feedbacklist: [],
    },
    {
      employee_id: 2,
      firstName: "Jack",
      lastName: "Shin",
      flag: "0",
      feedbacklist: [],
    },
  ],
};

const NoEmployees = {
  status: 1,
  message: "Success",
  employees: [],
};

const databaseError = {
  status: 0,
  message: "Database Error",
  employees: [],
};

describe("Employee List Component Test", () => {
  let wrapperEmployees;
  beforeEach(() => {
    wrapperEmployees = mount(
      <AppEmployeesProvider>
        <Employeelist props={employees} />
      </AppEmployeesProvider>
    );
  });

  afterEach(cleanup);
  it("Employee List Component Renders Correctly", () => {
    const div = document.createElement("div");
    render(
      <AppEmployeesProvider>
        <Employeelist props={employees} />
      </AppEmployeesProvider>,
      div
    );
  });

  it("Check if No assigned Employees", () => {
    const wrapper = mount(
      <AppEmployeesProvider>
        <Employeelist props={NoEmployees} />
      </AppEmployeesProvider>
    );
    const noEmployees = wrapper.find("li").exists();
    expect(noEmployees).toBeFalsy();
  });

  it("Check List of Employees", () => {
    const RowOnefirstName = wrapperEmployees.find("li span").at(0).text();
    const RowOneLastName = wrapperEmployees.find("li span").at(1).text();
    const RowtwofirstName = wrapperEmployees.find("li span").at(2).text();
    const RowtwoLastNames = wrapperEmployees.find("li span").at(3).text();
    expect(RowOnefirstName).toEqual("Mau");
    expect(RowOneLastName).toEqual("Maya");
    expect(RowtwofirstName).toEqual("Jack");
    expect(RowtwoLastNames).toEqual("Shin");
  });

  it("Check Database Error Display", () => {
    const wrapper = mount(
      <AppEmployeesProvider>
        <Employeelist props={databaseError} />
      </AppEmployeesProvider>
    );

    const errorDisplay = wrapper.find("div").text();
    expect(errorDisplay).toEqual("Database Error");
  });

  it("Check if Employee Component is Declare", () => {
    expect(wrapperEmployees.find(Employee).length).toBe(1);
  });

  it("Check if Employee Component is Declare", () => {
    expect(wrapperEmployees.find(FeedBackList).length).toBe(1);
  });

  it("Check if FeedBack Component is Diplayed When Button Click", () => {
    wrapperEmployees.find("button").at(0).simulate("click");
    expect(wrapperEmployees.find(".employee-container").exists()).toBeTruthy();
  });

  it("Check if FeedBack Component is Diplayed When Button Click", () => {
    wrapperEmployees.find("button").at(2).simulate("click");
    expect(wrapperEmployees.find(".employee-container").exists()).toBeTruthy();
  });

  it("Check if FeedBack Component is Diplayed When Button Click", () => {
    wrapperEmployees.find("button").at(3).simulate("click");
    expect(
      wrapperEmployees.find(".feedback-list-container").exists()
    ).toBeTruthy();
  });

  it("Assigned Employee List Component Test Check Matches Snapshot Display Employee", () => {
    const Snapshot = renderer
      .create(
        <AppEmployeesProvider>
          <Employeelist props={employees} />
        </AppEmployeesProvider>
      )
      .toJSON();
    expect(Snapshot).toMatchSnapshot();
  });

  it("Assigned Employee List Component Test Check Matches Snapshot Display Error", () => {
    const Snapshot = renderer
      .create(
        <AppEmployeesProvider>
          <Employeelist props={databaseError} />
        </AppEmployeesProvider>
      )
      .toJSON();
    expect(Snapshot).toMatchSnapshot();
  });
});
