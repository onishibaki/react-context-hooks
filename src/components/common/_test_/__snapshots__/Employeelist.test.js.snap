// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Assigned Employees Component Test Assigned Employee List Component Test Check Matches Snapshot Display Employee 1`] = `
<div
  className="main-container"
>
  <div
    className="sc-fzqBZW bwulSy"
  >
    <div
      className="sc-fzoyAV ghgnHc"
    >
      <button
        className="sc-fzoLag sc-fzoXzr bchcwt"
        onClick={[Function]}
      >
        Add
      </button>
      <span>
        <a
          href="/"
        >
          Home
        </a>
      </span>
    </div>
    <ul
      className="sc-fzpjYC funJtt"
    >
      <li>
        <div
          className="sc-fzqNJr byqwuN"
        >
          <label>
            First name: 
          </label>
          <span
            className="sc-fznxsB kxNAV"
          >
            Mau
          </span>
          <label>
            Last name: 
          </label>
          <span
            className="sc-fznxsB kxNAV"
          >
            Maya
          </span>
        </div>
        <button
          className="sc-fzoLag fjaKOX"
          onClick={[Function]}
        >
          Delete
        </button>
        <button
          className="sc-fzoLag fjaKOX"
          onClick={[Function]}
        >
          Update
        </button>
        <button
          className="sc-fzoLag fjaKOX"
          onClick={[Function]}
        >
          View
        </button>
      </li>
      <li>
        <div
          className="sc-fzqNJr byqwuN"
        >
          <label>
            First name: 
          </label>
          <span
            className="sc-fznxsB kxNAV"
          >
            Jack
          </span>
          <label>
            Last name: 
          </label>
          <span
            className="sc-fznxsB kxNAV"
          >
            Shin
          </span>
        </div>
        <button
          className="sc-fzoLag fjaKOX"
          onClick={[Function]}
        >
          Delete
        </button>
        <button
          className="sc-fzoLag fjaKOX"
          onClick={[Function]}
        >
          Update
        </button>
        <button
          className="sc-fzoLag fjaKOX"
          onClick={[Function]}
        >
          View
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Assigned Employees Component Test Assigned Employee List Component Test Check Matches Snapshot Display Error 1`] = `
<div
  className="main-container"
>
  Database Error
</div>
`;

exports[`Employee List Component Test Assigned Employee List Component Test Check Matches Snapshot Display Employee 1`] = `
<div
  className="main-container"
>
  <div
    className="sc-fzqBZW bwulSy"
  >
    <div
      className="sc-fzoyAV ghgnHc"
    >
      <button
        className="sc-fzoLag sc-fzoXzr bchcwt"
        onClick={[Function]}
      >
        Add
      </button>
      <span>
        <a
          href="/"
        >
          Home
        </a>
      </span>
    </div>
    <ul
      className="sc-fzpjYC funJtt"
    >
      <li>
        <div
          className="sc-fzqNJr byqwuN"
        >
          <label>
            First name: 
          </label>
          <span
            className="sc-fznxsB kxNAV"
          >
            Mau
          </span>
          <label>
            Last name: 
          </label>
          <span
            className="sc-fznxsB kxNAV"
          >
            Maya
          </span>
        </div>
        <button
          className="sc-fzoLag fjaKOX"
          onClick={[Function]}
        >
          Delete
        </button>
        <button
          className="sc-fzoLag fjaKOX"
          onClick={[Function]}
        >
          Update
        </button>
        <button
          className="sc-fzoLag fjaKOX"
          onClick={[Function]}
        >
          View
        </button>
      </li>
      <li>
        <div
          className="sc-fzqNJr byqwuN"
        >
          <label>
            First name: 
          </label>
          <span
            className="sc-fznxsB kxNAV"
          >
            Jack
          </span>
          <label>
            Last name: 
          </label>
          <span
            className="sc-fznxsB kxNAV"
          >
            Shin
          </span>
        </div>
        <button
          className="sc-fzoLag fjaKOX"
          onClick={[Function]}
        >
          Delete
        </button>
        <button
          className="sc-fzoLag fjaKOX"
          onClick={[Function]}
        >
          Update
        </button>
        <button
          className="sc-fzoLag fjaKOX"
          onClick={[Function]}
        >
          View
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Employee List Component Test Assigned Employee List Component Test Check Matches Snapshot Display Error 1`] = `
<div
  className="main-container"
>
  Database Error
</div>
`;
